crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -154,24 +168,36 @@ impl<S: States> Deref for State<S> {
 ///     next_game_state.set(GameState::InGame);
 /// }
 /// ```
-#[derive(Resource, Debug)]
+#[derive(Resource, Debug, Default)]
 #[cfg_attr(
     feature = "bevy_reflect",
     derive(bevy_reflect::Reflect),
-    reflect(Resource, Default)
+    reflect(Resource)
 )]
-pub struct NextState<S: States>(pub Option<S>);
-
-impl<S: States> Default for NextState<S> {
-    fn default() -> Self {
-        Self(None)
-    }
+pub enum NextState<S: States> {
+    /// No transition has been planned
+    #[default]
+    Unchanged,
+    /// There is a transition planned for state `S`
+    Set(S),
+    /// There is a planned removal of the state `S`
+    Remove,
 }
```

Comment:
This is actually something I disagree with - I don't want to force every state definition to include a built-in "None" for when they aren't relevant. There is already a rust structure for handling that, so I'd prefer to use it.

I can see the argument of not exposing the ability to have None states for user's directly edited states - but for these derived ones I very much feel it is necessary.

Reply:
To be clear - it can work technically, but I don't believe that would make a good DX, and it doesn't play nice with existing rust structures. For example - it'll get rid of the ability to use `?` and the default `Iter` implementation of Option when using states or in the `compute` function.

Reply:
The thing is, computed states can be `None` when they don't match a condition, but they require `NextState`.
So there is no need to add this removal feature to `NextState`.

I understand a scenario where
```rs
 AppState {
    #[default]
    Menu,
    InGame {
        paused: bool,
    },
}
```
an `InGamePause` computed state would make no sense during `AppState::Menu`, so it's better to remove it. It just wouldn't happen through `NextState`

Reply:
Ah - yes. So in essence it's tied to the changes to how NextState should be handled overall. I am still worried about the scope of changes required to move away from NextState - but for me this tips the scales so I will look into that more later today/tomorrow.